### PR TITLE
feat(dqlite): add IsController service on machine domain

### PR DESCRIPTION
**Pull request by cderici**
Adds the `IsController` service function in the machine domain.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```
TEST_PACKAGES="./domain/machine/... -count=1 -race" make run-go-tests
```

### CODE REVIEW

File: domain/machine/state/state.go [hunk 1]
```diff
@@ -193,6 +194,38 @@ func (st *State) GetMachineLife(ctx context.Context, machineName machine.Name) (
...
 
+// IsController returns whether the machine is a controller machine.
+// It returns a NotFound if the given machine doesn't exist.
+func (st *State) IsController(ctx context.Context, machineName machine.Name) (bool, error) {
+	db, err := st.DB()
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	machineNameParam := sqlair.M{"name": machineName}
```
Comment (manadart) [inline domain/machine/state/state.go:205]: Use the type you created for this.

**Review by nvinuesa — APPROVED** [review]
Code LGTM, thanks

File: domain/machine/state/state.go [hunk 2]
```diff
@@ -187,10 +187,42 @@ func (st *State) GetMachineLife(ctx context.Context, machineName machine.Name) (
...
+// IsController returns whether the machine is a controller machine.
+// It returns a NotFound if the given machine doesn't exist.
+func (st *State) IsController(ctx context.Context, mName machine.Name) (bool, error) {
+	db, err := st.DB()
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	machineNameParam := machineName{Name: mName}
+	query := `SELECT is_controller AS &machineIsController.is_controller FROM machine WHERE name = $machineName.name`
```
Comment (manadart) [inline domain/machine/state/state.go:205]: You can just write `SELECT &machineIsController.is_controller FROM ...`

File: domain/machine/state/state.go [hunk 3]
```diff
@@ -187,10 +187,42 @@ func (st *State) GetMachineLife(ctx context.Context, machineName machine.Name) (
...
+	queryStmt, err := st.Prepare(query, machineNameParam, machineIsController{})
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	var isController bool
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		result := machineIsController{}
```
Comment (manadart) [inline domain/machine/state/state.go:213]: This whole function is simple enough to declare this at the top, pass it to prepare, and return its inner bool without reassignment.

**Review by manadart — APPROVED** [review]
Just a few nits.

**Pull request merged**